Implement region-based state merging

futile opened the pull request:
This PR implements a state merging technique based on manually placed regions.  For this purpose we added two new functions, `klee_open_merge()` and `klee_close_merge()`. States which run into a `klee_close_merge()` are paused from scheduling, and merged once all states that branched from the same state which ran into `klee_open_merge()` have arrived at any `klee_close_merge()`.

New tests are also included in this PR, which give a good overview of how this technique can be used.

The main part of the functionality is implemented in the class `BoundedMergeHandler`. The `SpecialFunctionHandler` generates an instance of this class every time a state runs into a `klee_open_merge()` call.  This instance is appended to a `std::vector<klee::ref<BoundedMergeHandler>>` in the `ExecutionState` that passed the merge open point.  This stack is also copied during forks, and a stack is used to support nested merge regions.  Once a state runs into a `klee_close_merge()`, the `SpecialFunctionHandler` notifies the top `BoundedMergeHandler` of the state's stack, pauses the state from scheduling, and tries to merge it with all other states that already arrived at the same close merge point. This top instance then is popped from the stack, resulting in a decrease of the ref count of the `BoundedMergeHandler`. Since the only references to this `BoundedMergeHandler` are in the stacks of the `ExecutionState`s currently in the merging region, once the ref count reaches zero every state which ran into the same `klee_open_merge()` is now paused and waiting to be merged. The destructor of the `BoundedMergeHandler` then continues the scheduling of the corresponding paused states.

Because the random-path searcher was not able to halt scheduling of `ExecutionState`, we implemented a backtracking technique to randomly select an `ExecutionState` without scheduling paused states.

We also implemented a heuristic incomplete merging version, which will prematurely release states if the remaining states take much longer than the states which already have reached a corresponding `klee_close_merge()`.  For this heuristic, it is required that the BoundedMergingSearcher prioritizes states in a `BoundedMergeHandler`. This option is turned on with the `-incomplete-bounded-merge` flag.

We removed the `MergingSearcher` and `BumpMergingSearcher`, as their functionality was unusable in larger code because they only merged after every other state terminated.

ccadar commented:
Thanks for the contribution, Felix.  Besides getting the Travis CI tests to pass, in order to accept such a feature we would need to:
(1) break the single commit into smaller commits and multiple PRs which could be approved one at a time (e.g, one independent PR would be the removal of the old merging code).
(2) understand the impact of these changes on KLEE when this functionality is not used.  Ideally there should be no impact.
(3) understand why you chose not to extend the old merging functionality. 
(4) test the change on a larger regression test suite.  @andreamattavelli has a good infrastructure for this, but he won't be available to help during the next few weeks.
(5) document the feature on the KLEE website, so that others can use it.

delcypher commented:
@ccadar Part of the cause of the compilation failures is the use of C++11 in this PR. I would very much like to use C++11 in KLEE's code. Currently we are blocked from making the move (proposed in #314 quite a long time ago) due to LLVM 2.9. I'd like to call again for support for this version of LLVM to be removed (#318). LLVM 2.9 is 6 years old at this point and is irrelevant. I can't even compile this version of LLVM anymore with a modern gcc/clang.

corrodedHash commented:
@ccadar Thank you for the feeback!

**To (1)**:  
  It is very possible to seperate this into two pull requests, one for the 
  removal and one for the implementation, and we should be able to seperate
  the implementation into the merging itself, and the optional incomplete 
  merging.  
  Would this suffice? Do you require the pull requests themself to contain
  multiple commits?

**To (2)**:  
  This has to be checked, I will discuss this with my team.
  
**To (3)**:  
  The old merging technique was flawed, in that it only merged when all other 
  paths were terminated. In larger code bases this is not feasible, as paths
  were basically indefinitely stopped once they reached a `klee_merge()`.  
  Our technique allows paths to be merged and continued regardless of other
  paths' termination state.  
  This approach was very different from the already implemented functionality,
  which is why we chose not to extend the existing functionality.

  **To (4) and (5)**:
  Sure, I will see to this.

futile commented:
@ccadar 

Thank you for your feedback! We have now split this PR into three individual PRs:

* PR #689 removes the old `BumpMergingSearcher`. The PR explains our reasoning for the removal (instead of extension).
* PR #690 implements the basic version of our region-based state merging, and includes a description of the additional overhead.
* PR #691 complements the basic version with additional functionality to prevent waiting states from 'starving' in case other running states take too long to reach the merge point.

While we have added tests for the new functionality, and made sure the existing tests still pass, we are also currently running the new implementation against a testsuite consisting of the coreutils, binutils, diffutils, findutils, vim and python. We still have to match the results from these runs against runs of an unmodified KLEE, and make sure that we have not introduced any bugs.

Regarding documentation, we currently plan to write it up as soon as we can be sure that the basic approach is ok, and once we know which part of the new functionality will be merged. Is that okay? (We would be fine with blocking the PRs until documentation is ready, in case they are accepted.)

P.S.: The last current travis failure seems to be due to a spurious failure retrieving a software package, and unrelated to our changes. Anyway, this PR can probably be closed in favor of continuing discussion in #689, #690 and #691.

ccadar commented:
Hi @futile , that's great, thanks.  Indeed, that was a spurious failure; I restarted the job and it passed.  But let's close this PR and focus on the three individual PRs.  

I might take a bit until I review those PRs, as I'd first like to make another release of KLEE.
